From 8075b310607eb2f82adaf66a0d5a5340f38569e3 Mon Sep 17 00:00:00 2001
From: c4pp4
Date: Wed, 21 Oct 2026 12:47:20 +0200
Subject: [PATCH 1/1] Use libnotify when available (evolution-alarm-notify)

notify-osd in Unity does not support notification actions and shows
the reminder sent via GNotification (with the default action and the
"Reminders" button) as a dialog. Send a plain notification via
libnotify instead, but only when running in Unity.

Signed-off-by: c4pp4
---
 .../evolution-alarm-notify/CMakeLists.txt     |  9 +++++
 .../evolution-alarm-notify/e-alarm-notify.c   | 28 +++++++++++++++
 .../evolution-alarm-notify.c                  | 35 +++++++++++++++++++
 3 files changed, 72 insertions(+)

diff --git a/src/services/evolution-alarm-notify/CMakeLists.txt b/src/services/evolution-alarm-notify/CMakeLists.txt
index 5939990..ca9c1ba 100644
--- a/src/services/evolution-alarm-notify/CMakeLists.txt
+++ b/src/services/evolution-alarm-notify/CMakeLists.txt
@@ -33,6 +33,15 @@ target_compile_definitions(evolution-alarm-notify PRIVATE
 	-DLOCALEDIR=\"${LOCALE_INSTALL_DIR}\"
 )
 
+pkg_check_modules(LIBNOTIFY libnotify)
+
+if(LIBNOTIFY_FOUND)
+	target_compile_definitions(evolution-alarm-notify PRIVATE -DHAVE_LIBNOTIFY)
+	target_compile_options(evolution-alarm-notify PUBLIC ${LIBNOTIFY_CFLAGS})
+	target_include_directories(evolution-alarm-notify PUBLIC ${LIBNOTIFY_INCLUDE_DIRS})
+	target_link_libraries(evolution-alarm-notify ${LIBNOTIFY_LDFLAGS})
+endif()
+
 target_compile_options(evolution-alarm-notify PUBLIC
 	${CANBERRA_CFLAGS}
 	${DATA_SERVER_CFLAGS}
diff --git a/src/services/evolution-alarm-notify/e-alarm-notify.c b/src/services/evolution-alarm-notify/e-alarm-notify.c
index dc24419..e1dc4d0 100644
--- a/src/services/evolution-alarm-notify/e-alarm-notify.c
+++ b/src/services/evolution-alarm-notify/e-alarm-notify.c
@@ -32,6 +32,9 @@
 #include "libedataserverui/libedataserverui.h"
 
 #include "e-alarm-notify.h"
+#ifdef HAVE_LIBNOTIFY
+#include <libnotify/notify.h>
+#endif
 
 struct _EAlarmNotifyPrivate {
 	ESourceRegistry *registry;
@@ -310,6 +313,23 @@ e_alarm_notify_build_notif_id (const EReminderData *rd)
 	return g_string_free (string, FALSE);
 }
 
+#ifdef HAVE_LIBNOTIFY
+static void
+e_alarm_notify_libnotify (const gchar *summary,
+			  const gchar *body)
+{
+	NotifyNotification *n;
+
+	if (!notify_is_initted ())
+		return;
+
+	n = notify_notification_new (summary, body, "appointment-soon");
+	notify_notification_set_timeout (n, NOTIFY_EXPIRES_DEFAULT);
+	notify_notification_show (n, NULL);
+	g_object_unref (n);
+}
+#endif
+
 static gboolean
 e_alarm_notify_display (EAlarmNotify *an,
 			const EReminderData *rd,
@@ -327,6 +347,11 @@ e_alarm_notify_display (EAlarmNotify *an,
 
 	if (g_settings_get_boolean (an->priv->settings, "notify-enable-display") &&
 	    !g_hash_table_contains (an->priv->notification_ids, notif_id)) {
+#ifdef HAVE_LIBNOTIFY
+		if (notify_is_initted ()) {
+			e_alarm_notify_libnotify (_("Reminders"), description);
+		} else {
+#endif
 		GNotification *notification;
 		GIcon *icon;
 		gchar *detailed_action;
@@ -348,6 +373,9 @@ e_alarm_notify_display (EAlarmNotify *an,
 		g_application_send_notification (G_APPLICATION (an), notif_id, notification);
 
 		g_object_unref (notification);
+#ifdef HAVE_LIBNOTIFY
+		}
+#endif
 
 		g_hash_table_insert (an->priv->notification_ids, notif_id, NULL);
 	} else {
diff --git a/src/services/evolution-alarm-notify/evolution-alarm-notify.c b/src/services/evolution-alarm-notify/evolution-alarm-notify.c
index 555b44d..56ffc07 100644
--- a/src/services/evolution-alarm-notify/evolution-alarm-notify.c
+++ b/src/services/evolution-alarm-notify/evolution-alarm-notify.c
@@ -26,6 +26,30 @@
 
 #include "e-alarm-notify.h"
 
+#ifdef HAVE_LIBNOTIFY
+#include <libnotify/notify.h>
+
+static gboolean
+is_running_unity (void)
+{
+	const gchar *desktop;
+	gchar **desktops;
+	gboolean found = FALSE;
+	gint ii;
+
+	desktop = g_getenv ("XDG_CURRENT_DESKTOP");
+	if (!desktop)
+		return FALSE;
+
+	desktops = g_strsplit (desktop, ":", -1);
+	for (ii = 0; desktops[ii] && !found; ii++)
+		found = g_strcmp0 (desktops[ii], "Unity") == 0;
+	g_strfreev (desktops);
+
+	return found;
+}
+#endif
+
 #ifdef G_OS_UNIX
 #include <glib-unix.h>
 
@@ -62,6 +86,12 @@ main (gint argc,
 
 	gtk_init (&argc, &argv);
 
+#ifdef HAVE_LIBNOTIFY
+	/* notify-osd in Unity shows notifications with actions as dialogs */
+	if (is_running_unity ())
+		notify_init ("Evolution Alarm Notify");
+#endif
+
 	e_xml_initialize_in_main ();
 	_libedataserverui_init_icon_theme ();
 
@@ -83,5 +113,10 @@ main (gint argc,
 
 	g_object_unref (alarm_notify);
 
+#ifdef HAVE_LIBNOTIFY
+	if (notify_is_initted ())
+		notify_uninit ();
+#endif
+
 	return exit_status;
 }
-- 
2.55.0

